### PR TITLE
feat: Support rest params in function calls

### LINKED ISSUE #377: Support rest parameters
I hope that variable parameters supported.

When I compiled the below codes.
```typescript
function sum(...args: i32[]): i32 {
    let total: i32 = 0;
    for (let index = 0; index < args.length; index ++) {
        total += args[index];
    }
    return total;

}
assert(sum(1, 2, 3) == 6);
```
I got the errors.
```
ERROR TS2554: Expected 1 arguments, but got 3.
```

btw, the [online studio](https://webassembly.studio/?f=mmzz9venspr) was unavailable.  
```
ERROR: Entry file 'main.ts' not found.
```

### LINKED ISSUE #2291: No support for rest paramaters
I have the following code:

```as.ts add
export function add(...args: Uint32Array): i32 {
  return args.reduce((a, b) => a + b, 0)
}
```

I'm getting error "AS215: Optional parameter must have an initializer.":

image

When I try to add initializer:

```as.ts add
export function add(...args = []: Uint32Array): i32 {
  return args.reduce((a, b) => a + b, 0)
}
```

Then I'm getting error "TS1048: A rest parameter cannot have an initializer":

image

_Originally posted by @jerrygreen in https://github.com/AssemblyScript/assemblyscript/issues/377#issuecomment-810021811_

### PR DESCRIPTION
**Pull request by mattjohnsonpint**
Fixes #377
Fixes #2291

Changes proposed in this pull request:

- Implement support for [rest parameters](https://www.typescriptlang.org/docs/handbook/2/functions.html#rest-parameters) when calling functions or constructors.

Note, does not yet support _rest arguments_, as the spread operator and iterators would need to be implemented first.

ie. this now works:

```ts
function foo<T>(...args: T[]): void {
}

foo(1, 2, 3, 4, 5);
```

but this does not (yet):

```ts
const nums = [1, 2, 3, 4, 5];
foo(...nums);
```

## Behavior

Fundamentally,

```ts
function foo<T>(...args: T[]): void { }
foo(1, 2, 3);
```

is equivalent to:

```ts
function foo<T>(args: T[] = []): void { }
foo([1, 2, 3]);
```

The only difference is that the caller does not construct the array - but the array still exists in the function scope.

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -1075,7 +1075,7 @@ export class Signature {
     }
 
     // check rest parameter
-    if (this.hasRest != target.hasRest) return false; // TODO
```
Comment (CountBleck) [inline src/types.ts:1078]: I think you should keep the TODO.

`(...args: i32[]) => void` should be compatible with `() => void` (relatively easy to support) and `(a: i32, b: i32) => void` (hard) at some point in the future.

File: src/types.ts [hunk 2]
```diff
@@ -1087,7 +1087,7 @@ export class Signature {
     let thisParameterTypes = this.parameterTypes;
     let targetParameterTypes = target.parameterTypes;
     let numParameters = thisParameterTypes.length;
-    if (numParameters != targetParameterTypes.length) return false; // TODO
```
Comment (CountBleck) [inline src/types.ts:1090]: Same as above.

File: tests/compiler/call-rest-err.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  }
+  return sum;
+}
+
+function count<T>(...args: T[]): i32 {
+  return args.length;
+}
+
+sum('a', 'b'); // expect a type mismatch error on each argument
+count(1, 'a'); // expect a type mismatch error on the second argument
```
Comment (MaxGraey) [inline tests/compiler/call-rest-err.ts:14]: Can you also test failure case when rest params placed not at the end? For example:
```ts
function wrongOrder<T>(...args: T[], extra: i32): i32 {
  return args.length + extra;
}
wrongOrder('a', 1); // expect a rest parameter must be last in a parameter list
```
Error:
```
A rest parameter must be last in a parameter list. (1014)
```

Comment (MaxGraey) [inline tests/compiler/call-rest-err.ts:14]: Also need a case with zero rest arguments:
```ts
sum(); // Expected at least 1 arguments, but got 0.(2555)
```

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: Will do.

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: That should be allowed, no?

Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: No, at least on typescript it's not allowed

**Comment by MaxGraey** [comment]
There is also another very specific case when arguments are processed not in a loop:
```ts
function fixedArgs(...args: i32[]): i32 {
  return args[0] + args[1];
}

fixedArgs(1, 2); // ok
```
But this is become undefined behavior for such cases:
```ts
fixedArgs(1);
fixedArgs(1, 2, 3);
```
TypeScript does not generate an error in this case, but we must anticipate this option and reject all options except those with two arguments to the call site

File: tests/compiler/call-rest-err.ts (hunk 3 above)
Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: It is though:
https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAZxAWwBQDocEMBOA5sgFyJjoBGApvgNoC6AlGRWjfogN4BQiiAG2pQU6RAF5EABgDcfRMDicMQkTAnSANIgDWGgsSxCwhKAAsZidQB5dlgNT2YTbvP6o0ie5IPI6MBjl+AF95fGEQfCQPOVCeCARkOCEjOEIMDwwARk0AJk0AZiYXAHoSxAA2eMTk6lT0zOLEMukgA

Various tutorials describe it as such also:

https://www.typescripttutorial.net/typescript-tutorial/typescript-rest-parameters/

> A rest parameter allows a function to accept zero or more arguments of the specified type.

https://www.tutorialsteacher.com/typescript/rest-parameters

> ... During the second function call, we do not pass any arguments as the rest parameters. This is accepted by the compiler ...

Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: I'm afraid you'll have to analyze the CFG for that. However, it will have to be done for `fixedArgs` case (see comment below) anyway... In general, it's not rest params it's not such a trivial thing just because of such edge cases

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: > I'm afraid you'll have to analyze the CFG for that.

Not sure what you mean.  `sum()` does currently work.  It returns `0` as expected.

Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: > Not sure what you mean. sum() does currently work. It returns 0 as expected.

I mean this:
```ts
function getFirst(...args: i32[]): i32 {
   return args.length ? args[0] : 0;
}
```
or
```ts
function sum(...args: i32[]): i32 {
   let res = 0;
   for (let i = 0; i < args.length; i++) {
      res += args[i];
   }
   return res;
}
```
In both cases `getFirst()` and `sum()` will work due to presenting default or accumulator locals. But if you slighly modify these functions it stop to work:
```ts
function getFirst(...args: i32[]): i32 {
   return args[0];
}
getFirst(); // ERROR
```

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: Sure.  I would expect those to error, due to index out of range. The `args` array is initialized to `[]` so, `args[0]` should give index out of range.  In regular TypeScript it returns `undefined`.

But isn't that just a factor of how arrays work in AssemblyScript vs TypeScript?  I'm not sure how you would expect it to behave differently?

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: To clarify, I am create an empty array literal as the initializer for rest parameters.

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: > ... however, `count()` should error since it's defined as `count<T>` and there are no arguments to infer the type from. It currently crashes the compiler instead of erroring correctly. I'll fix that shortly.

I fixed the crash and added a test for this case.

Not sure what else you'd like me to do about the rest.  From my perspective it seems to be working well.

**Comment by CountBleck** [comment]
@MaxGraey that kind of checking would probably be a TODO, since I don't believe the compiler in its current state has a way to note metadata like that on functions. Maybe that would be set aside for a separate pass, in combination with a custom IR (or something that could detect things like `const aaarghs = args`)?

In any case, the heuristic wouldn't work as soon as `someFunction(args)` is in the function (`someFunction` could mutate the array).

File: src/compiler.ts [hunk 4]
```diff
@@ -6477,7 +6503,11 @@ export class Compiler extends DiagnosticEmitter {
       let declaration = originalParameterDeclarations[minArguments + i];
       let initializer = declaration.initializer;
       let initExpr: ExpressionRef;
-      if (initializer) {
+      if (declaration.parameterKind === ParameterKind.Rest) {
+        const arrExpr = new ArrayLiteralExpression([], declaration.range.atEnd);       
+        initExpr = this.compileArrayLiteral(arrExpr, type, Constraints.ConvExplicit);
```
Comment (MaxGraey) [inline src/compiler.ts:6508]: Is it possible to use `compileStaticArrayLiteral` here?

**Comment by mattjohnsonpint** [comment]
> ... TypeScript does not generate an error in this case

It doesn't generate an error for `fixedArgs(1)`, because `args[1]` is `undefined`, and `1 + undefined === NaN`, so it returns `NaN`.  But your other example `fixedArgs(1, 2, 3)` should be valid in both TS and AS.  There's no requirement that all args passed are used.

> ... but we must anticipate this option and reject all options except those with two arguments to the call site and preferably during compile time

Why?  I don't think we do that type of analysis for anything similar such as passing arrays.

IMHO `function fixedArgs(...args: i32[]): i32` and `function fixedArgs(args: i32[]=[]): i32` should be identical in how the args are treated within the function.  It's only the caller that changes syntax.

File: src/compiler.ts (hunk 4 above)
Comment (mattjohnsonpint) [reply src/compiler.ts:6508]: Good question.  I will attempt it and let you know! 👍

Comment (CountBleck) [reply src/compiler.ts:6508]: Reading `compileArrayLiteral`'s code, it looks like the user can opt into `StaticArray` by declaring `...args: StaticArray<i32>`. @mattjohnsonpint maybe you should make a test for that...

Anyhow, the user might desire the admittedly strange case where the function might mutate and play around with the `args` array (or perhaps more commonly, use it to initialize a class field?).

Comment (mattjohnsonpint) [reply src/compiler.ts:6508]: Nope.  It doesn't work.  It would be the equivalent of doing this:

```ts
function foo<T>(args: T[] = new StaticArray<T>())
```

and a static array can't be assigned to a regular array.

File: tests/compiler/call-rest-err.ts (hunk 3 above)
Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: No, in JS, Go, Rust, [Python](https://godbolt.org/z/1Pxqf9hfa) and other langs rest params passed thought stack. That's why max size of rest args is 64k in JS.

Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: I updated the PR description to clarify this point.

Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: Ideally, it would be the same to use shadow stack. Especially the infrastructure is already in place for this, as it is necessary for GC:
https://github.com/AssemblyScript/assemblyscript/blob/main/src/passes/shadowstack.ts

**Comment by MaxGraey** [comment]
Why I'm concerned about the implementation via a dynamic array. Many math methods such as `Math.min`, `Math.max`, `Math.hypot` as well as Array#push, Array#concat and many others require this feature. If every time there will be created an implicit array even for one argument, it will significantly affect the performance and also significantly increase memory pressure.

File: tests/compiler/call-rest-err.ts (hunk 3 above)
Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: Hmmm.. I hadn't seen the Python detail.  There it does seem to be using the stack as you said.  I think this might be possible due to Python's dynamic nature - but I'm no expert here.  The shadow stack sounds interesting, but that is beyond my capabilities, I'm afraid.  Would you consider allowing this as-is and then making that a TODO for future perf improvement?

FWIW, I searched through several authoritative sources and could not find the max args limit you mentioned.  TypeScript just passes `...` on to JavaScript, so it doesn't impose any limit.  JavaScript in the MDN docs or the ECMAScript spec doesn't mention a limit (that I could find).  The one bit of detail that seemed relevant was in the MDN docs [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters#the_difference_between_rest_parameters_and_the_arguments_object):

> The `arguments` object is not a real array, while rest parameters are `Array` instances, meaning methods like `sort()`, `map()`, `forEach()` or `pop()` can be applied on it directly.

... not that we're talking about the JS `arguments` option here, but the description of rest parameters being actual array instances seems like something we'd have portability or user-expectation issues with if AssemblyScript's were not.

**Comment by mattjohnsonpint** [comment]
> ... it will significantly affect the performance and also significantly increase memory pressure.

Those are valid concerns indeed.  I'm open to suggestions.

Would it help if the array were `@unmanaged` ?

File: tests/compiler/call-rest-err.ts (hunk 3 above)
Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: The limit is depending on VM's stack for each browser:
https://stackoverflow.com/a/22747272

**Comment by mattjohnsonpint** [comment]
> Would it help if the array were `@unmanaged` ?

though it couldn't be implemented as easily as this:

```ts
@unmanaged
class UnmanagedArray<T> extends Array<T> { }
```

... because: `ERROR AS207: Unmanaged classes cannot extend managed classes and vice-versa.`


Another thought.  Does AssemblyScript have any concept (internal or otherwise) similar to `stackalloc` in C# (or Go's implicit stack allocations, etc.)?

**Comment by mattjohnsonpint** [comment]
Also, I'm thinking about your concern with `Math.max`, etc.  In typescript, both rest params and rest args are supported, so you can do something like this:

```ts
function maxNum(str: string): number {
  const arr = str.split(' ').map(s => parseInt(s))
  return Math.max(...arr)
}

console.log(maxNum("10 3 40 1 2")) // 40
console.log(maxNum("100 200 300")) // 300
```

While this PR doesn't implement rest args, I presumed that we'd want to do that eventually.

Since in this example, the array size cannot be known at compile time, I'm really not sure how one could implement `Math.max` that would allow this, without using a dynamically sized array.

**Comment by MaxGraey** [comment]
Well, you will definitely need to separate the static and dynamic case for rest parameters. Perhaps it would be a good option to leave everything as it is and in the future add a special optimizing pass or optimization for the case when all the arguments are known at compile time and not to use an array for this case but to use shadow stack or static data segments

**Comment by mattjohnsonpint** [comment]
> Well, you will definitely need to separate the static and dynamic case for rest parameters. Perhaps it would be a good option to leave everything as it is and in the future add a special optimizing pass or conditional optimization for the case when all the arguments are known at compile time and not to use an array for this case but to use shadow stack or static data segments

That sounds reasonable.  I could also see an approach similar to generics, where the compiler makes internal representations of the function signature with fixed sizes to be used by those callers.  ie, `Math.max$2(p0, p1)`, `Math.max$3(p0, p1, p2)`, etc.

So then, for now - you good with this PR as is?

File: tests/compiler/call-rest-err.ts (hunk 3 above)
Comment (mattjohnsonpint) [reply tests/compiler/call-rest-err.ts:14]: Good to know.  Thanks for the link.

Comment (MaxGraey) [reply tests/compiler/call-rest-err.ts:14]: Hmm, I don't see an expected error in stderr for `1014`

**Review by MaxGraey — APPROVED** [review]

**Comment by MaxGraey** [comment]
Let someone else approve this before merging

**Comment by MaxGraey** [comment]
> but this does not (yet):
> ```ts
> const nums = [1, 2, 3, 4, 5];
> foo(...nums);
> ```

This is not relate to rest params. It's spread syntax which use iterator under hood.

I have changed my opinion about approving. The point is that I don't see the point of using a dynamic array for variadic arguments. Nobody does that, it's not efficient and semantically incorrect. Rest params is an ArrayLike readonly object that only has a length and can read elements by index. If you represent it as a full dynamic array, you will leak methods like `push` and `pop`:

> The arguments object is not a real array, while rest parameters are Array instances, meaning methods like sort(), map(), forEach() or pop() can be applied on it directly.

[Source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters#the_difference_between_rest_parameters_and_the_arguments_object)

So we can't use the array here

**Review by HerrCai0907 — APPROVED** [review]
LGTM

**Comment by CountBleck** [comment]
Okay, I just tested it myself as a sanity check, and everything looks good...I'll merge it now!
